### ISSUE TITLE
When carla is installed in /usr/lib64, carla-rack and carla-patchbay doesn't work.

### ISSUE THREAD

**Issue by ycollet** [LMMS/lmms#2429]
Carla can be installed in /usr/lib64.
But a hard link to /usr/lib instead of /usr/lib64 is still in carla.cpp:

```
CarlaInstrument::CarlaInstrument(InstrumentTrack* const instrumentTrack, const Descriptor* const descriptor, const bool isPatchbay)
    : Instrument(instrumentTrack, descriptor),
      kIsPatchbay(isPatchbay),
      fHandle(NULL),
      fDescriptor(isPatchbay ? carla_get_native_patchbay_plugin() : carla_get_native_rack_plugin()),
      fMidiEventCount(0)
{
    fHost.handle      = this;
    fHost.uiName      = NULL;
    fHost.uiParentId  = 0;

    // figure out prefix from dll filename
    QString dllName(carla_get_library_filename());

#if defined(CARLA_OS_LINUX)
    fHost.resourceDir = strdup(QString(dllName.split("/lib/carla")[0] + "/share/carla/resources/").toUtf8().constData());
#else
    fHost.resourceDir = NULL;
#endif
```

This line should be adapted to lib or lib64:
fHost.resourceDir = strdup(QString(dllName.split("/lib/carla")[0] + "/share/carla/resources/").toUtf8().constData());

**Comment by tresf** [comment]
@ycollet thanks.  Tagging @falkTX as he's the author of this plugin.

**Comment by falkTX** [comment]
Is there a LIB_PREFIX macro or something used by lmms?
It can be used there instead of hardcoded '/lib/'.

**Comment by tresf** [comment]
@falkTX no, not that I'm aware of.  LADSPA discovery looks like this currently: [`src/core/LadspaManager.cpp#L52`](https://github.com/LMMS/lmms/blob/master/src/core/LadspaManager.cpp#L52)

This may eventually be a good candidate for @Lukas-W's [`PluginFactory`](https://github.com/LMMS/lmms/blob/master/src/core/PluginFactory.cpp).

**Comment by falkTX** [comment]
I'm pretty sure something related to CMAKE_INSTALL_PREFIX is passed to the code as a C define.
But I don't really know which one exactly.

**Comment by tresf** [comment]
> I'm pretty sure something related to CMAKE_INSTALL_PREFIX is passed to the code as a C define. But I don't really know which one exactly.

[`LIBDIR`](https://cmake.org/cmake/help/git-master/module/GNUInstallDirs.html) perhaps via [`src/lmmsconfig.h.in`](https://github.com/LMMS/lmms/blob/master/src/lmmsconfig.h.in)?

**Comment by tresf** [comment]
@falkTX I've taken a second look at this and I think @ycollet's investigation is accurate... that the bug lies in the hard-coding of `/lib/carla`.  The cmake may help here, but they're still unnecessary if the prefix is already known in `dllName`.  The `/lib/carla` split seems only meant to isolate the `/usr` prefix, which is more part of the system's installed Carla version and less part of any LMMS or cmake build configuration.

A theoretical fix could be to use the following logic, assuming relative paths are allowed... 

```diff
- fHost.resourceDir = strdup(QString(dllName.split("/lib/carla")[0] + "/share/carla/resources/").toUtf8().constData());
+ fHost.resourceDir = strdup(QString(dllName.split("/carla/")[0] + "../share/carla/resources/").toUtf8().constData());
```

**Comment by falkTX** [comment]
Using relative paths is for sure the right way to go.
(regarding using compile-time prefix, we have to remember carla can be installed on a different prefix, so we can't use this)
Since some distros use /lib64 instead of /lib, splitting using '/lib' is not very good.

I did some tests here, I recommend to use this:
```
-    fHost.resourceDir = strdup(QString(dllName.split("/lib/carla")[0] + "/share/carla/resources/").toUtf8().constData());
+    dllName = dllName.section("/carla/", 0, dllName.count("/carla/")-1) + "/../share/carla/resources/";
+    dllName = QFileInfo(dllName).absoluteFilePath();
+    fHost.resourceDir = strdup(dllName.toUtf8().constData());
```

This needs QFileInfo added as include on the top of the file.

I can do a PR against the stable-1.3 branch if you prefer.

**Comment by tresf** [comment]
@falkTX, `/lib` handles both.  Take a second look, it matches `/lib64` too, we're grabbing before it, not after.

**Comment by falkTX** [comment]
Yes, though it will fail if there's a double /lib folder somewhere, as it's grabbing the first one.
Say a path like `/opt/libs/carla-appimage/lib/carla/libcarla_standalone2.so`.

The code I propose is more safe, as it uses the last `/carla/` string found. It will work even if there's more than one that matches "/carla/".

**Comment by tresf** [comment]
> ```diff
> -    fHost.resourceDir = strdup(QString(dllName.split("/lib/carla")[0] + "/share/carla/resources/").toUtf8().constData());
> +    dllName = dllName.section("/carla/", 0, dllName.count("/carla/")-1) + "/../share/carla/resources/";
> +    dllName = QFileInfo(dllName).absoluteFilePath();
> +    fHost.resourceDir = strdup(dllName.toUtf8().constData());
> ```

Thanks.

If we're introducing `QFileInfo`, why not just... 

```cpp
QString dllName(carla_get_library_filename());
QDir path(QFileInfo(dllName).dir());
path.cdUp();
fHost.resourceDir = strdup(path.absolutePath().toUtf8().constData());
```

Instead of relying on string data, it will grab the grandparent of the shared object.

**Comment by falkTX** [comment]
That's even better :)
I'm so used to work with raw strings that I sometimes forget this.

**Comment by tresf** [comment]
Closed via d711b8b.  This fix is available in the `stable-1.2` branch.  It will not exist in master until the `stable-1.2` changes have been synced into master.
